neqo-transport/src/fc.rs

```diff
@@ -161,9 +164,131 @@ impl SenderFlowControl<StreamId> {
     }
 }
 
+#[derive(Debug)]
+pub struct ReceiverFlowControl<T>
+where
+    T: Debug + Sized,
+{
+    /// The thing that we're counting for.
+    subject: T,
+    /// Max amount of outstanding data
+    max_wnd: u64,
+    // Last max data sent.
+    max_data: u64,
+    // Retired bytes.
+    retired: u64,
+    frame_pending: bool,
+}
+
+impl<T> ReceiverFlowControl<T>
+where
+    T: Debug + Sized,
+{
+    /// Make a new instance with the initial value and subject.
+    pub fn new(subject: T, max_bytes: u64) -> Self {
+        Self {
+            subject,
+            max_wnd: max_bytes,
+            max_data: max_bytes,
+            retired: 0,
+            frame_pending: false,
+        }
+    }
+
+    /// Check if received data exceeds the allowed flow control limit.
+    pub fn check_allowed(&self, new_end: u64) -> bool {
+        new_end <= self.max_data
+    }
+
+    /// Some data has been read, retired them and maybe send flow control
+    /// update.
+    pub fn retired(&mut self, retired: u64) {
+        if retired <= self.retired {
+            return;
+        }
+
+        self.retired = retired;
+        let maybe_new_max = self.retired + self.max_wnd;
+        if maybe_new_max > (self.max_wnd / 2) + self.max_data {
+            self.max_data = maybe_new_max;
+            self.frame_pending = true;
+        }
```

Comment:
So this is the same code as in send_flowc_update(), with an extra variable.  Maybe you can add an internal function that takes a value (x) where `x = self.max_wnd / 2` or `x = 0` as appropriate.

However, I think that we can make this better by not tracking max_data in quite the same way here.  That is, rather than only update the maximum when we see a need to do so (we retire more than half the window or we receive a blocked frame) instead, we can update the value at the time that the frame is sent.

That is, this function will just work out if an update might be needed.  That is, if we are retiring/reading data, we would set frame_pending to true if retired is within buffer_size/2 of the current maximum; or if we receive a blocked frame, we update if retired is less than buffer_size from the current maximum.  Then, when we are asked to produce a frame we calculate the new maximum as retired + buffer_size and pass the value to max_data_sent(), which clears frame_pending and sets max_data to that value.

```suggestion
        if self.retired + self.max_active / 2 > self.max_data {
            self.frame_pending = true;
        }
```

The effect of this should be that multiple reads will result in sending the largest MAX_DATA we can.  And, if we lose the packet containing MAX_DATA, we will send an even bigger value in the next packet if there were more reads in the meantime.